Calibration/HcalCalibAlgos/src/SimAnalyzerMinbias.cc

```diff
@@ -34,72 +34,89 @@
-  
-  // Root tree members
-  int   mydet, mysubd, depth, iphi, ieta, cells;
-  float mom0_MB, mom1_MB, mom2_MB, mom3_MB, mom4_MB;
-  struct myInfo{
-    double theMB0, theMB1, theMB2, theMB3, theMB4;
-    void MyInfo() {
-      theMB0 = theMB1 = theMB2 = theMB3 = theMB4 = 0;
-    }
-  };
-  std::map<HcalDetId,myInfo>               myMap;
-  edm::EDGetTokenT<edm::HepMCProduct>      tok_evt_;
-  edm::EDGetTokenT<edm::PCaloHitContainer> tok_hcal_;
+  std::map<HcalDetId,HcalSimMinbias::myInfo> myMap_;
+  edm::EDGetTokenT<edm::HepMCProduct>        tok_evt_;
+  edm::EDGetTokenT<edm::PCaloHitContainer>   tok_hcal_;
 };
 
 // constructors and destructor
 
-SimAnalyzerMinbias::SimAnalyzerMinbias(const edm::ParameterSet& iConfig) {
+SimAnalyzerMinbias::SimAnalyzerMinbias(const edm::ParameterSet& iConfig, 
+				       const HcalSimMinbias::Counters* count) {
+
   // get name of output file with histogramms
-  fOutputFileName = iConfig.getUntrackedParameter<std::string>("HistOutFile", "simOutput.root");
   timeCut         = iConfig.getUntrackedParameter<double>("TimeCut", 500);
     
   // get token names of modules, producing object collections
   tok_evt_ = consumes<edm::HepMCProduct>(edm::InputTag("generator"));
   tok_hcal_ = consumes<edm::PCaloHitContainer>(edm::InputTag("g4SimHits","HcalHits"));
 
-#ifdef debugLog
-  std::cout << "Use Time cut of " << timeCut << " ns and store o/p in "
-	    << fOutputFileName << std::endl;
-#endif
-}
-  
-SimAnalyzerMinbias::~SimAnalyzerMinbias() {
-  // do anything here that needs to be done at desctruction time
-  // (e.g. close files, deallocate resources etc.)
-}
-  
-void SimAnalyzerMinbias::beginRun( const edm::Run& r, const edm::EventSetup& iSetup) {
+  edm::LogInfo("AnalyzerMB") << "Use Time cut of " << timeCut << " ns";
+  myMap_.clear();
 }
   
-void SimAnalyzerMinbias::endRun( const edm::Run& r, const edm::EventSetup& iSetup) {
+SimAnalyzerMinbias::~SimAnalyzerMinbias() { }
+
+void SimAnalyzerMinbias::endStream() {
+
+  for (std::map<HcalDetId,HcalSimMinbias::myInfo>::const_iterator itr=myMap_.begin(); itr != myMap_.end(); ++itr) {
+    HcalDetId id = itr->first;
+    std::map<HcalDetId,HcalSimMinbias::myInfo>::iterator itr1 = (globalCache()->myMap_).find(id);
+    if (itr1 == globalCache()->myMap_.end()) {
+      HcalSimMinbias::myInfo info;
+      globalCache()->myMap_[id] = info;
+      itr1 = (globalCache()->myMap_).find(id);
+    }
+    itr1->second.theMB0 += itr->second.theMB0;
+    itr1->second.theMB1 += itr->second.theMB1;
+    itr1->second.theMB2 += itr->second.theMB2;
+    itr1->second.theMB3 += itr->second.theMB3;
+    itr1->second.theMB4 += itr->second.theMB4;
+  }
 }
   
-void SimAnalyzerMinbias::beginJob() {
-  hOutputFile   = new TFile( fOutputFileName.c_str(), "RECREATE" ) ;
-
-  myTree = new TTree("SimJet","SimJet Tree");
+void SimAnalyzerMinbias::globalEndJob(const HcalSimMinbias::Counters* count) {
+   
+  TFile* hOutputFile = new TFile(count->fOutputFileName_.c_str(), "RECREATE") ;
```

Comment:
This is a memory leak. Please change so instead of ```new``` you just create the ```TFile``` on the stack.